src/fasta.rs

```diff
@@ -1,209 +1,932 @@
+    ///     comment,
+    ///     vec![
+    ///         FastaRecord {
+    ///             header: "a".to_string(),
+    ///             contents: "sequence".to_string(),
+    ///             line_range: (2, 4),        
+    ///         }
+    ///     ]
+    /// );
+    ///
+    /// let comment = FastaParser::<String>::new(
+    ///     FastaParseSettings::new().allow_preceding_comment(false)
+    /// ).parse_str("comment\n>a\nsequence").unwrap();
+    /// assert_eq!(
+    ///     comment,
+    ///     vec![
+    ///         FastaRecord {
+    ///             header: "".to_string(),
+    ///             contents: "comment".to_string(),
+    ///             line_range: (1, 2),
+    ///         },
+    ///         FastaRecord {
+    ///             header: "a".to_string(),
+    ///             contents: "sequence".to_string(),
+    ///             line_range: (2, 4),        
+    ///         }
+    ///     ]
+    /// );
+    /// ```
+    allow_preceding_comment: bool,
+}
 
-        let mut lines: Vec<String> = vec![];
-        let mut title: Option<String> = None;
+// "Builder-lite" pattern: https://matklad.github.io/2022/05/29/builder-lite.html
+impl FastaParseSettings {
+    /// Initializes settings "strictly" (See [`Self::strict()`]).
+    pub fn new() -> Self {
+        Self::strict()
+    }
 
-        let handle = BufReader::new(handle);
+    /// Initializes settings "strictly":
+    /// * [`Self::concatenate_headers`]: false
+    /// * [`Self::allow_preceding_comment`]: false
+    pub fn strict() -> Self {
+        Self {
+            concatenate_headers: false,
+            allow_preceding_comment: false,
+        }
+    }
 
-        for line in handle.lines() {
-            let line = line?;
+    /// Initializes settings "laxly":
+    /// * [`Self::concatenate_headers`]: true
+    /// * [`Self::allow_preceding_comment`]: true
+    pub fn lax() -> Self {
+        Self {
+            concatenate_headers: true,
+            allow_preceding_comment: true,
+        }
+    }
+
+    /// Changes the setting for [`Self::concatenate_headers`]
+    pub fn concatenate_headers(mut self, concatenate_headers: bool) -> Self {
+        self.concatenate_headers = concatenate_headers;
+        self
+    }
+
+    /// Changes the setting for [`Self::allow_preceding_comment`]
+    pub fn allow_preceding_comment(mut self, allow_preceding_comment: bool) -> Self {
+        self.allow_preceding_comment = allow_preceding_comment;
+        self
+    }
+}
+
+impl Default for FastaParseSettings {
+    fn default() -> Self {
+        Self::new()
+    }
+}
 
-            if let Some(ch) = line.chars().next() {
-                if ch == '>' {
-                    if let Some(existing_title) = title {
-                        result.push((existing_title, lines.join("")));
-                        lines.clear();
-                    }
+enum ParserState {
+    StartOfFile {
+        /// only non-empty if settings.allow_preceding_comment is false
+        contents: String,
```

Comment:
my string concatenation speed alarms went off here, but actually I think this is basically fine. Might be slightly faster to accumulate, like, a Vec<&str> or something and then join them at the end, but :shrug:

Reply:
Yeah I thought about that as well, but I *think* the quadratic behavior won't manifest here. Maybe I should add 10,000 line comment, concatenated header, and sequence body checks to be safe :-)

Reply:
Also I don't think we can hold a `Vec<&str>` because lifetimes™, we get chunks of data fed into us from a BufRead instead of a single string we can hold onto an reference chunks out of :-(

Reply:
Yeah, you're right :/

Reply:
(I agree re not quadratic, since Strings in Rust are mutable)

Reply:
Added some tests for quadratic-ness in most recent commit, using a `panic_after` helper so they'll panic if the runtime ever grows significantly.